### PR TITLE
Translate relative links to absolute links in docs

### PR DESCRIPTION
**Pull request by Atish-iaf**

### CODE REVIEW

**Review by antoninbas — COMMENTED** [review]
I initially thought you would target specific directories for link substitution (`build/`, and potentially others). Your solution (looking for links with an extension) could work, but you have to be careful:
- ideally some extensions may need to be ignored (images?)
- we need to handle all links IMO (all relative and absolute links): `build/`, `../build/`, `../../build/`, `../../../build/`, `/build/`. You shouldn't limit yourselves to just 2 patterns, and ideally you would have a single generic regex expression (or maybe 2).

Finally, I suggest that you re-generate the latest version of the docs (`content/docs/v1.11.0`) using your updated script, and that you check-in the changes as part of this PR. This will have 2 benefits: 1) we can evaluate whether your changes to the script are correct before merging the PR, and 2) it will fix up the v1.11 version of the website documentation.

**Comment by Atish-iaf** [comment]
Thanks @antoninbas for valuable suggestions . Please help to review again.

File: content/docs/v1.11.0/docs/antctl.md [hunk 1]
```diff
@@ -656,7 +656,7 @@ testmulticast-vw7gx5b9 test3-sender-1               0       10
 
 ### Showing memberlist state
 
-`antctl` agent command `get memberlist` (or `get ml`) print the state of memberlist cluster of Antrea Agent.
+`antctl` agent command `get memberlist` (or `get ml`) prints the state of memberlist cluster of Antrea Agent.
```
Comment (antoninbas) [inline content/docs/v1.11.0/docs/antctl.md:659]: while this is not a big deal, it means that when you generated the docs for v1.11.0, you used the wrong version of antrea (you checked out main, instead of the v1.11.0 tag). You may want to do it again with the correct tag, to avoid confusion.

File: scripts/pkg/update-docs.go [hunk 2]
```diff
@@ -179,13 +179,34 @@ func fixupMarkdownFile(path string) error {
 	if DryRun {
 		return nil
 	}
+
+	// Prefix link to be added before relative links in
+	// order to translate relative links into absolute links.
+	prefix := fmt.Sprintf("(https://github.com/antrea-io/antrea/blob/%s/", version)
+
+	// Translate relative links to directories build, pkg, hack, ci, CHANGELOG and LICENSE file into absolute links.
+	relativeLinksExp := regexp.MustCompile(`\([\./]*(build/|pkg/|hack/|ci/|CHANGELOG|LICENSE)+[A-Za-z0-9\.\-_/]*\)`)
```
Comment (antoninbas) [inline scripts/pkg/update-docs.go:188]: yes, this is pretty much what I had in mind
a couple of suggestions:
- can we move the list of file prefixes to a separate variable and build the regex using that variable? it will make maintenance easier. `const filePrefixes = []string{"build/",...}`
- is the `+` at the end of the prefix list correct?

File: scripts/pkg/update-docs.go [hunk 3]
```diff
@@ -179,13 +179,34 @@ func fixupMarkdownFile(path string) error {
...
+	relativeLinks := relativeLinksExp.FindAll(md, -1)
+	for _, rl := range relativeLinks {
+		var start int
+		existingLink := string(rl)
+		for i, char := range existingLink {
+			if char != '(' && char != '.' && char != '/' {
+				start = i
+				break
+			}
+		}
```
Comment (antoninbas) [inline scripts/pkg/update-docs.go:198]: I think you should use a capturing group in your regex to avoid having to do this

Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: done, PTAL

File: scripts/pkg/update-docs.go (hunk 2 above)
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: - Slice or const array is not supported in Go, so did a workaround by creating a function.
- No, it was a mistake, removed. Thanks!

File: content/docs/v1.11.0/docs/antctl.md (hunk 1 above)
Comment (Atish-iaf) [reply content/docs/v1.11.0/docs/antctl.md]: done.

File: scripts/pkg/update-docs.go [hunk 4]
```diff
@@ -179,13 +185,28 @@ func fixupMarkdownFile(path string) error {
...
+
+	// Translate relative links to directories and files returned by getFilePrefixes() into absolute links.
+	relativeLinksExp := regexp.MustCompile(`\([./]*((` + strings.Join(getFilePrefixes(), `|`) + `)[\w.\-/]*\))`)
+	relativeLinks := relativeLinksExp.FindAll(md, -1)
+	for _, rl := range relativeLinks {
+		submatch := relativeLinksExp.FindSubmatch(rl)
+		suffix := string(submatch[1])
+		absLink := fmt.Sprintf("%s%s", prefix, suffix)
+		md = []byte(strings.Replace(string(md), string(rl), absLink, -1))
+	}
```
Comment (antoninbas) [inline scripts/pkg/update-docs.go:201]: I was suggesting doing it in one step with something like:
```go
relativeLinksExp := regexp.MustCompile(`\([./]*((build/|ci/|CHANGELOG)[\w.\-/]*)\)`)
relativeLinksExp.ReplaceAllString(md, "(https://github.com/antrea-io/antrea/blob/${1})")
```
see https://regex101.com/r/tlT2if/1 for the regex in action

Capture group 1 (`${1}`) is the part of the link that needs to be appended to the prefix.

File: scripts/pkg/update-docs.go [hunk 5]
```diff
@@ -40,6 +40,12 @@ var (
 	DryRun      bool
 )
 
+func getFilePrefixes() []string {
+	// All Relative links present in .md files inside content/docs/<version>/ to following
+	//directories and files will be translated into their appropriate absolute links.
+	return []string{"build/", "pkg/", "hack/", "ci/", "CHANGELOG", "LICENSE"}
+}
```
Comment (antoninbas) [inline scripts/pkg/update-docs.go:47]: probably better to stick to a variable in the `fixupMarkdownFile` function:
```go
filePrefixes := []string{"build/", "pkg/", "hack/", "ci/", "CHANGELOG", "LICENSE"}
```

otherwise the function name (`getFilePrefixes`) doesn't provide enough context

File: scripts/pkg/update-docs.go (hunk 4 above)
Comment (antoninbas) [reply scripts/pkg/update-docs.go:201]: actually I have been thinking about this and I think we need a solution that's a bit more robust.

Imagine we have a sub-folder called `build/` somewhere in the antrea repo hierarchy (which is different from `/build`). Then the current logic may not work.
What we need instead:
- find all relative links (e.g., `../build/charts/antrea/crds/supportbundlecollection.yaml`)
- translate them into a relative link starting at the root of the directory (e.g., `../build/charts/antrea/crds/supportbundlecollection.yaml` => `build/charts/antrea/crds/supportbundlecollection.yaml`)
- prepend the absolute http link prefix

The second step is what is currently missing.

IMO, something like this may work well:
```go
re := <regex that matches all markdown links, excluding the ones that are URLs (start with `proto://`)>
mdDir := <relative path of the md file directory with respect to the root of the directory (e.g., 'docs' for 'docs/helm.md')>
re.ReplaceAllStringFunc(md, func(link string) string {
    newLink := link
    newLink = strings.TrimLeft(newLink, "(")
    newLink = strings.TrimRight(newLink, ")")
    // using path is more appropriate than path/filepath here
    if !path.IsAbs(newLink) {
        newLink = path.Join(mdDir, newLink)
    }
    newLink = path.Clean(newLink)
    for _, prefix := range prefixes {
        if strings.HasPrefix(newLink, prefix) {
            return fmt.Sprintf("(https://github.com/antrea-io/antrea/blob/%s/%s)", version, newLink)
        }
    }
    return link
})
```

Some variations to that code may be needed. This has also made me realize that we need to start adding unit tests for that code, as this is a pretty "complex" operation.

Comment (antoninbas) [reply scripts/pkg/update-docs.go:201]: That's a good point, and I had not thought about that. However, I think it's better to go with the more future-proof approach now. We could imagine having `docs/ci` at some point in the future. With the new approach, we will avoid bad surprises down the road. Please consider adding unit tests for these edge cases.

File: scripts/pkg/update-docs.go (hunk 5 above)
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: done.

File: content/docs/v1.11.0/docs/maintainers/release.md [hunk 6]
```diff
@@ -23,7 +23,7 @@ release. We use `<TAG>` as a placeholder for the release tag (e.g. `v1.4.0`).
      message must be *exactly* `"Update CHANGELOG for <TAG> release"`, as a bot
      will look for this commit and cherry-pick it to update the main branch
      (starting with Antrea v1.0). The
-     [process-changelog.go](../../hack/release/process-changelog.go) script may
+     [process-changelog.go](https://github.com/antrea-io/antrea/blob/v1.11.0/hack/release/process-changelog.go) script may
      be used to easily generate links to PRs and the Github profiles of PR
      authors.
    - a commit to update [VERSION](../../VERSION) as needed, using the following
```
Comment (Atish-iaf) [inline content/docs/v1.11.0/docs/maintainers/release.md:29]: @antoninbas could you please help to find why this is not getting translated. I have added unit test for the same, it gets translated there.
Same issue at L85

File: scripts/pkg/update-docs.go (hunk 4 above)
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: done, PTAL.

File: content/docs/v1.11.0/docs/maintainers/release.md (hunk 6 above)
Comment (Atish-iaf) [reply content/docs/v1.11.0/docs/maintainers/release.md:29]: Please ignore, I didn't run `make` command after making changes.

File: scripts/pkg/update-docs_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+			expected: []byte("texts[filename](https://github.com/antrea-io/antrea/blob/v1.11.0/build/yamls/filename.yml)texts"),
+		},
+		{
+			fullPath: "website/content/docs/v1.11.0/docs/maintainers/release.md",
+			md:       []byte("texts[VERSION](../../VERSION)texts"),
+			expected: []byte("texts[VERSION](https://github.com/antrea-io/antrea/blob/v1.11.0/VERSION)texts"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
```
Comment (antoninbas) [inline scripts/pkg/update-docs_test.go:73]: could you try giving an explicit name to each of your subtests?

**Review by antoninbas — COMMENTED** [review]
you should update the Makefile with a target to run unit tests (see other antrea-io repositories) and update .github/workflows/go.yml to run the tests as part of CI.

File: scripts/pkg/update-docs.go [hunk 8]
```diff
@@ -165,7 +174,29 @@ func syncDirs(sourceDocsPath string, destDocsPath string, docDir *DocDir) error
 	return nil
 }
 
-func fixupMarkdownFile(path string) error {
+func translateRelativeLinks(fullPath, version string, md *[]byte) {
+	re := regexp.MustCompile(`\]\(([^http][\/\.\-\w\d]+)\)`)
+	mdDir := strings.TrimPrefix(fullPath, fmt.Sprintf("website/content/docs/%s/", version))
```
[the remark this answers is not in the document]
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: yes, done.

File: scripts/pkg/update-docs_test.go (hunk 7 above)
Comment (Atish-iaf) [reply scripts/pkg/update-docs_test.go]: added, but names may not be good enough!

File: scripts/pkg/update-docs.go [hunk 9]
```diff
@@ -40,6 +41,14 @@ var (
 	DryRun      bool
 )
 
+func getFilePrefixes() []string {
```
[the remark this answers is not in the document]
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: done.

File: scripts/pkg/update-docs.go [hunk 10]
```diff
@@ -165,7 +174,29 @@ func syncDirs(sourceDocsPath string, destDocsPath string, docDir *DocDir) error
 	return nil
 }
 
-func fixupMarkdownFile(path string) error {
+func translateRelativeLinks(fullPath, version string, md *[]byte) {
+	re := regexp.MustCompile(`\]\(([^http][\/\.\-\w\d]+)\)`)
```
[the remark this answers is not in the document]
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: thanks for reminding.

File: scripts/pkg/update-docs.go [hunk 11]
```diff
@@ -165,7 +174,29 @@ func syncDirs(sourceDocsPath string, destDocsPath string, docDir *DocDir) error
 	return nil
 }
 
-func fixupMarkdownFile(path string) error {
+func translateRelativeLinks(fullPath, version string, md *[]byte) {
```
[the remark this answers is not in the document]
Comment (Atish-iaf) [reply scripts/pkg/update-docs.go]: done

**Comment by Atish-iaf** [comment]
> you should update the Makefile with a target to run unit tests (see other antrea-io repositories) and update .github/workflows/go.yml to run the tests as part of CI.

Done, do you think it is OK to upload unit test coverage report to Codecov ? Or should we skip uploading report as of now?
Thanks

File: .github/workflows/go.yml [hunk 12]
```diff
@@ -43,3 +43,47 @@ jobs:
     - name: Build
       working-directory: ./scripts
       run: make golangci
+
+  check-changes:
```
Comment (antoninbas) [inline .github/workflows/go.yml:47]: we don't need this optimization here

File: .github/workflows/go.yml [hunk 13]
```diff
@@ -43,3 +43,47 @@ jobs:
...
+  test-unit:
+    needs: check-changes
+    if: ${{ needs.check-changes.outputs.has_changes == 'yes' }}
+    name: Unit test
+    runs-on: [ubuntu-latest]
+    steps:
+    - name: Check-out code
+      uses: actions/checkout@v2
+    - name: Set up Go using version from go.mod
+      uses: actions/setup-go@v2
```
Comment (antoninbas) [inline .github/workflows/go.yml:70]: update to the latest version of these actions: `actions/checkout@v3` and `actions/setup-go@v4`.
In particular `actions/setup-go@v4` supports caching out of the box:
```yaml
    - name: Set up Go using version from go.mod
      uses: actions/setup-go@v4
      with:
        go-version-file: 'go.mod'
        cache-dependency-path: 'scripts/go.sum'
```
That means you can get rid of the custom caching logic

File: scripts/pkg/update-docs_test.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+		expected []byte
+	}{
+		{
+			name:     "no translation-absolute external link",
+			fullPath: "website/content/docs/v1.11.0/somefile.md",
+			content:  []byte("texts[filename](https://github.com/antrea-io/antrea/blob/main/test/test.yaml)texts"),
+			expected: []byte("texts[filename](https://github.com/antrea-io/antrea/blob/main/test/test.yaml)texts"),
+		},
+		{
+			name:     "no translation-relative link to a file which is synced",
```
Comment (antoninbas) [inline scripts/pkg/update-docs_test.go:40]: I am not sure what "synced" means. I assume it means "translated into a website doc page"? Since you use "synced" to describe multiple tests, maybe add a comment to define the term, just above the `tests` list definition.

**Review by antoninbas — COMMENTED** [review]
couple of small comments, otherwise LGTM

File: .github/workflows/go.yml (hunk 13 above)
Comment (antoninbas) [reply .github/workflows/go.yml:70]: Please update these 2 actions for all existing workflows, not just the new one

File: scripts/pkg/update-docs_test.go (hunk 14 above)
Comment (Atish-iaf) [reply scripts/pkg/update-docs_test.go:40]: Here, 'synced' means that the file is present in website repository as it has been copied(synced) from Antrea repository by `func syncDirs` while updating docs.

Comment (antoninbas) [reply scripts/pkg/update-docs_test.go:40]: please add a comment to this effect

**Review by antoninbas — APPROVED** [review]
